File: Binary_Tree/vertical_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<tuple<int, int, int> > res;
    void dfs(TreeNode* node, int x, int y){
        if(node == NULL)
            return;
        res.push_back(make_tuple(x, -y, node -> val));
        dfs(node -> left, x - 1, y - 1);
        dfs(node -> right, x + 1, y - 1);
    }
    bool hasSameXCord(int i, int j){
        return (get<0>(res[i]) == get<0>(res[j]));
    }
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        dfs(root, 0, 0);
        sort(res.begin(), res.end());
        int sz = res.size();
        vector<vector<int> > ans;
        int indx = -1;
        for(int i = 0; i < sz; i++){
            if(i == 0 || (i > 0 && !hasSameXCord(i, i - 1))){
                vector<int> temp;
                ans.push_back(temp);
                indx++;
            }
            ans[indx].push_back(get<2>(res[i]));
        }
        return ans;
    }
};

File: Binary_Tree/vertical_order_traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#            3 (0,0)

# 9(-1, -1)                         20 (1, -1)
#                       15(0, -2)              7(2, -2)

# 9   (-1,-1,9)
# 3   (0,0,3)
# 15  (0, -2, 15)
# 20  (1, -1, 20)
# 7   (2, -2, 7)

## x - ascending
## y - descending
## val - ascending 


class Solution(object):
    def __init__(self):
        self.tuple_list = []
    
    def dfs(self, node, x, y):
        if node is None:
            return
        self.tuple_list.append((x, -y, node.val))
        self.dfs(node.left, x - 1, y - 1)
        self.dfs(node.right, x + 1, y - 1)
        
        
    def verticalTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        self.dfs(root, 0, 0)
        self.tuple_list.sort()
        sz = len(self.tuple_list)
        ans = []
        temp = []
        for i in range(0, sz):
            if i > 0 and not self.tuple_list[i-1][0] == self.tuple_list[i][0]:
                ans.append(temp)
                temp = []
            temp.append(self.tuple_list[i][2])
        ans.append(temp)
        return ans